pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -395,6 +402,18 @@ protected void finishOperation() {
     }
   }
 
+  /**
+   * When TTL is enabled for upsert, this function is used to remove expired keys from the primary key indexes.
+   */
+  @Override
+  public void removeExpiredPrimaryKeys(Comparable watermark) {
+    if (_upsertTTLConfig.getTtlInMs() > 0) {
+      doRemoveExpiredPrimaryKeys((Long) watermark);
```

Comment:
We want to check the `_stopped` flag, and also use `startOperation()` and `finishOperation()` to prevent getting interrupted. You may refer to how other operations are handled

Reply:
Gotcha, thanks for the review!